text/0000-buildpack-input-vars.md

```diff
@@ -0,0 +1,142 @@
+# Meta
+[meta]: #meta
+- Name: Replace positional args to Buildpack executables with env vars
+- Start Date: 2021-11-18
+- Author(s): [jkutner](https://github.com/jkutner)
+- Status: Draft Draft
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal to replace the positional arguments of Buildpack executables with named environment variables.
+
+# Definitions
+[definitions]: #definitions
+
+- *positional arguments* - values passed into an executable command such that they are accessible as `$1`, `$2`, etc
+
+# Motivation
+[motivation]: #motivation
+
+Postional arguments to Buildpack executables have been the way for the buildpack execution engine to provide inputs for buildpacks since buildpacks v1 was created. However, positional arguments are limiting, and what they represent is not obvious without reading the spec.
+
+In the spec today they are defined as:
+
+* `/bin/detect <platform[AR]> <plan[E]>`
+* `/bin/build <layers[EIC]> <platform[AR]> <plan[ER]>`
+
+Using env vars helps make these inputs more easily accesible from language bindings like [libcnb.bash](https://github.com/jkutner/libcnb.bash).
+
+# What it is
+[what-it-is]: #what-it-is
+
+We will deprecate the positional arguments to `bin/detect` and `bin/build` with the following environment variables:
+
+### bin/detect
+
+* `CNB_PLATFORM_DIR` - replaces the first positional argument to `bin/build`. Uses the same env var name as the Platform spec.
+* `CNB_PLAN_PATH` - replaces the second positional argument to `bin/build`. Uses the same env var name as the Platform spec.
+
+### bin/build
+
+* `CNB_BP_LAYERS_DIR` - replaces the first positional argument to `bin/build`. Uses `_BP_` to differentiate it from the `CNB_LAYERS_DIR` in the Plaform spec, which is a different value.
```

Comment:
I don't like that we're sacrificing a common env var name in the buildpack spec for a rarely-used env var in the platform spec. How do we feel about renaming the platform one?

Otherwise, buildpack authors will wonder why it's not `CNB_BP_PLAN_PATH`, etc.

Reply:
Does this require buildpack/platform APIs to change in lock step are we ok with the overlap?

Reply:
The lifecycle should be able to handle the overlapping case, so I think it's up to the team's discretion.

Reply:
> Does this require buildpack/platform APIs to change in lock step are we ok with the overlap?

Even if you release them both simultaneously most people will experience an overlap anyways as new buildpacks will run on old platforms.

Reply:
@sclevine any preference on the new platform env var name?

Reply:
`CNB_PARENT_LAYERS_DIR`